# <center> KPMG optimising the quality of customer datasets </center>

### Here is the background information on your task
Sprocket Central Pty Ltd , a medium size bikes & cycling accessories organisation, has approached Tony Smith (Partner) in KPMG’s Lighthouse & Innovation Team. 

Sprocket Central Pty Ltd  is keen to learn more about KPMG’s expertise in its Analytics, Information & Modelling team. 

Smith discusses KPMG’s expertise in this space (you can read more here). In particular, he speaks about how the team can effectively analyse the datasets to help Sprocket Central Pty Ltd grow its business.

Primarily, Sprocket Central Pty Ltd needs help with its customer and transaction data. 

The organisation has a large dataset relating to its customers, but their team is unsure how to effectively analyse it to help optimise its marketing strategy. 

However, in order to support the analysis, you spoke to the Associate Director for some ideas and she advised that <b>“the importance of optimising the quality of customer datasets cannot be underestimated. The better the quality of the dataset, the better chance you will be able to use it to drive company growth.” </b>

The client provided KPMG with 3 datasets:
- Customer Demographic 
- Customer Addresses
- Transactions data in the past 3 months

You decide to start the preliminary data exploration and identify ways to improve the quality of Sprocket Central Pty Ltd’s data.


### What you'll do
<b>Draft an email to your client identifying data quality issues and strategies to mitigate the issues</b>

You will start working with clients. Who are mid-size companies selling bicycles and accessories. They’re looking to expand into a new market, and we will assist them by showcasing our data analytic skills to recommend a suitable marketing strategy. 
Over the course of the client project, There’ll be three main tasks involved. 
- data quality assessment. 
- Being a data model build
- Visualise these results of analysis which you could present to the client.


In [22]:
import pandas as pd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
# make sure all column are display while we use head function 
pd.options.display.max_columns=None

# Excel Sheet have muliple sheets and Sheet names: Transactions , NewCustomerList, 
# CustomerDemographic, CustomerAddress


## Data Pepration Trasactions Dataset
## -------

In [23]:
# read data set Transactions
df_T =pd.read_excel(r'KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='Transactions')
df_T.head(50)

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                      transaction_id                                                                                                                              
1                                                   1                                                                                                                              
2                                                   2                                                                                                                              
3                                                   3                                                                                                                              
4                                                   4                                                                                                                              
5                                                   5                                                                                                                              
6                                                   6                                                                                                                              
7                                                   7                                                                                                                              
8                                                   8                                                                                                                              
9                                                   9                                                                                                                              
10                                                 10                                                                                                                              
11                                                 11                                                                                                                              
12                                                 12                                                                                                                              
13                                                 13                                                                                                                              
14                                                 14                                                                                                                              
15                                                 15                                                                                                                              
16                                                 16                                                                                                                              
17                                                 17                                                                                                                              
18                                                 18                                                                                                                              
19                                                 19                                                                                                                              
20                                                 20                                                                                                                              
21                                         

### 1- Observation Transactions dataset:
- What's the difference between list_price standrd_cost?
- What's the meaning of product_first Sold date?
- There are 13 NaN colums in dataset -> My recomandation to drop this columns

In [24]:
#change header for transaction dataset
h1 =df_T.iloc[0]
df_T = df_T[1:]
df_T.columns= h1
df_T.head(20)

0  transaction_id product_id customer_id     transaction_date online_order  \
1               1          2        2950  2017-02-25 00:00:00        False   
2               2          3        3120  2017-05-21 00:00:00         True   
3               3         37         402  2017-10-16 00:00:00        False   
4               4         88        3135  2017-08-31 00:00:00        False   
5               5         78         787  2017-10-01 00:00:00         True   
6               6         25        2339  2017-03-08 00:00:00         True   
7               7         22        1542  2017-04-21 00:00:00         True   
8               8         15        2459  2017-07-15 00:00:00        False   
9               9         67        1305  2017-08-10 00:00:00        False   
10             10         12        3262  2017-08-30 00:00:00         True   
11             11          5        1986  2017-01-17 00:00:00        False   
12             12         61        2783  2017-01-05 00:00:00         True   
13             13         35        1243  2017-02-26 00:00:00         True   
14             14         16        2717  2017-09-10 00:00:00        False   
15             15         12         247  2017-06-11 00:00:00        False   
16             16          3        2961  2017-10-10 00:00:00        False   
17             17         79        2426  2017-04-03 00:00:00        False   
18             18         33        1842  2017-06-02 00:00:00        False   
19             19         54        2268  2017-04-06 00:00:00         True   
20             20         25        3002  2017-01-28 00:00:00         True   

0  order_status           brand product_line product_class product_size  \
1      Approved           Solex     Standard        medium       medium   
2      Approved   Trek Bicycles     Standard        medium        large   
3      Approved      OHM Cycles     Standard           low       medium   
4      Approved  Norco Bicycles     Standard        medium       medium   
5      Approved  Giant Bicycles     Standard        medium        large   
6      Approved  Giant Bicycles         Road        medium       medium   
7      Approved        WeareA2B     Standard        medium       medium   
8      Approved        WeareA2B     Standard        medium       medium   
9      Approved           Solex     Standard        medium        large   
10     Approved        WeareA2B     Standard        medium       medium   
11     Approved   Trek Bicycles     Mountain           low       medium   
12     Approved      OHM Cycles     Standard           low       medium   
13     Approved   Trek Bicycles     Standard           low       medium   
14     Approved  Norco Bicycles     Standard          high        small   
15     Approved  Giant Bicycles     Standard        medium        large   
16     Approved   Trek Bicycles     Standard        medium        large   
17     Approved  Norco Bicycles     Standard        medium       medium   
18     Approved  Giant Bicycles     Standard        medium        small   
19     Approved        WeareA2B     Standard        medium       medium   
20     Approved  Giant Bicycles         Road        medium       medium   

0  list_price standard_cost product_first_sold_date  NaN  NaN  NaN  NaN  NaN  \
1       71.49         53.62                   41245  NaN  NaN  NaN  NaN  NaN   
2     2091.47        388.92                   41701  NaN  NaN  NaN  NaN  NaN   
3     1793.43        248.82                   36361  NaN  NaN  NaN  NaN  NaN   
4     1198.46         381.1                   36145  NaN  NaN  NaN  NaN  NaN   
5      1765.3        709.48                   42226  NaN  NaN  NaN  NaN  NaN   
6     1538.99        829.65                   39031  NaN  NaN  NaN  NaN  NaN   
7       60.34         45.26                   34165  NaN  NaN  NaN  NaN  NaN   
8     1292.84         13.44                   39915  NaN  NaN  NaN  NaN  NaN   
9     1071.23        380.74                   33455  NaN  NaN  Na

In [25]:
# information about the data type formate for table
df_T.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 1 to 20000
Data columns (total 26 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   transaction_id           20000 non-null  object 
 1   product_id               20000 non-null  object 
 2   customer_id              20000 non-null  object 
 3   transaction_date         20000 non-null  object 
 4   online_order             19640 non-null  object 
 5   order_status             20000 non-null  object 
 6   brand                    19803 non-null  object 
 7   product_line             19803 non-null  object 
 8   product_class            19803 non-null  object 
 9   product_size             19803 non-null  object 
 10  list_price               20000 non-null  object 
 11  standard_cost            19803 non-null  object 
 12  product_first_sold_date  19803 non-null  object 
 13  nan                      0 non-null      float64
 14  nan                   

### 2- Obsevation Transaction Table in data type
- we need to change dataype for transaction_id, product_id, customer_id to be int
- Make datatime here consistance formate first and change datatype transaction_date to be date
- online_order change data type to be bool 
- order_status, brand, product_line change, product_class, product_size will change data type to be category
- list_price, standard_cost will change them to be float  
- what's the data type for this column( product_first_sold_date)
- Delete Nan colums because will effect data analysis

In [26]:
# let's know how many rows 
len(df_T.index)

20000

# # Data Clean Transaction dataset
## ----------------------
- Handel missing Values
- Convert column values
- Changing datatypes
- Handel duplication
- Convert new columns by combing other columns
- Drop unnecessary columns


In [27]:
# we will know how many missing values in that data set
for i in df_T.columns :
    if df_T[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_T[i].isna().sum())

Null Values in  online_order Column : 360
Null Values in  brand Column : 197
Null Values in  product_line Column : 197
Null Values in  product_class Column : 197
Null Values in  product_size Column : 197
Null Values in  standard_cost Column : 197
Null Values in  product_first_sold_date Column : 197
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000
Null Values in  nan Column : 20000


### 3- Obsevation Transaction Table in Missing data
- What's your prefrance to handel missing data in the following columns 
 - 1- online_order Column : 360
 - 2- brand Column : 197
 - 3- product_line Column : 197
 - 4- product_class Column : 197
 - 5- product_size Column : 197
 - 6- standard_cost Column : 197
 - 7- product_first_sold_date Column : 197

### 1-Handel missing data
### ----
We need to understand context of data to make dicision regarding how to handel Nan values data.

In [28]:
# Let's know how many rows
len(df_T.index)

20000

In [29]:
# 1- observation online_order column has 360 rows null
# Let's know more about online_order data
df_T['online_order'].unique()
# let's confirm df after drop 4 rows 
#len(df.index)

array([False, True, nan], dtype=object)

In [30]:
# group data and try to findout majority
df_T.groupby('online_order').size()

online_order
False    9811
True     9829
dtype: int64

#### I can't impute based on majority. My suggestion here to replace nan value with False.


In [31]:
# we will replace 360 null rows to False
df_T['online_order'].fillna(False,inplace = True)
# let's confirm df_T update
df_T['online_order'].unique()

array([False,  True])

#### ----

In [32]:
# 2- observation brand column has 197 rows null
# let's know more about data
df_T['brand'].unique()

array(['Solex', 'Trek Bicycles', 'OHM Cycles', 'Norco Bicycles',
       'Giant Bicycles', 'WeareA2B', nan], dtype=object)

### We can't predict the missing value so, I will drop it. Because it's not significant 

In [33]:
df_T =df_T.dropna(subset = ['brand'])
df_T['brand'].unique()

array(['Solex', 'Trek Bicycles', 'OHM Cycles', 'Norco Bicycles',
       'Giant Bicycles', 'WeareA2B'], dtype=object)

### ----

In [34]:
# 3- observation product_line column has 197 rows null
# Let's know more about data
df_T['product_line'].unique()

array(['Standard', 'Road', 'Mountain', 'Touring'], dtype=object)

### I didn't find sign for nan data in brand column so I will run check null data again to make sure data set is clean

In [35]:
# we will know how many missing values in that data set
for i in df_T.columns :
    if df_T[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_T[i].isna().sum())

Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803
Null Values in  nan Column : 19803


In [39]:
# will drop the all colums with nan values 
df_T1= df_T.dropna(how="any", axis=1)

#Check dataset 
df_T1.head(50)

0  transaction_id product_id customer_id     transaction_date  online_order  \
1               1          2        2950  2017-02-25 00:00:00         False   
2               2          3        3120  2017-05-21 00:00:00          True   
3               3         37         402  2017-10-16 00:00:00         False   
4               4         88        3135  2017-08-31 00:00:00         False   
5               5         78         787  2017-10-01 00:00:00          True   
6               6         25        2339  2017-03-08 00:00:00          True   
7               7         22        1542  2017-04-21 00:00:00          True   
8               8         15        2459  2017-07-15 00:00:00         False   
9               9         67        1305  2017-08-10 00:00:00         False   
10             10         12        3262  2017-08-30 00:00:00          True   
11             11          5        1986  2017-01-17 00:00:00         False   
12             12         61        2783  2017-01-05 00:00:00          True   
13             13         35        1243  2017-02-26 00:00:00          True   
14             14         16        2717  2017-09-10 00:00:00         False   
15             15         12         247  2017-06-11 00:00:00         False   
16             16          3        2961  2017-10-10 00:00:00         False   
17             17         79        2426  2017-04-03 00:00:00         False   
18             18         33        1842  2017-06-02 00:00:00         False   
19             19         54        2268  2017-04-06 00:00:00          True   
20             20         25        3002  2017-01-28 00:00:00          True   
21             21         27        1582  2017-10-09 00:00:00         False   
22             22         37         595  2017-06-29 00:00:00          True   
23             23         37        2001  2017-04-08 00:00:00          True   
24             24         82         515  2017-10-18 00:00:00         False   
25             25         89        2822  2017-06-11 00:00:00         False   
26             26         64        2596  2017-01-10 00:00:00         False   
27             27         64        2666  2017-04-11 00:00:00          True   
28             28         19          76  2017-12-23 00:00:00         False   
29             29         72        3368  2017-10-13 00:00:00          True   
30             30         91        1173  2017-03-15 00:00:00         False   
31             31         88        2810  2017-09-05 00:00:00          True   
32             32          1        3214  2017-02-18 00:00:00         False   
33             33         25        1272  2017-03-20 00:00:00         False   
34             34         99        2003  2017-02-28 00:00:00          True   
35             35          0        2171  2017-08-20 00:00:00         False   
36             36         92        3356  2017-07-07 00:00:00         False   
37             37         14        1040  2017-01-09 00:00:00         False   
38             38          2        2916  2017-12-06 00:00:00         False   
39             39         12         427  2017-09-12 00:00:00         False   
40             40          0        2448  2017-11-28 00:00:00          True   
41             41         44        2570  2017-05-08 00:00:00          True   
42             42         79        3398  2017-05-14 00:00:00         False   
43             43         76         488  2017-03-17 00:00:00         False   
44             44         46         165  2017-12-22 00:00:00         False   
45             45         55        2787  2017-06-07 00:00:00          True   
46             46         66        3053  2017-06-20 00:00:00         False   
47             47         35         475  2017-05-07 00:00:00          True   
48             48         81        1305  2017-06-10 00:00:00          True   
49             49         86        1251  2017-06-17 00:00:00         False   
50             50         32        3103  2017-05-

In [40]:
df_T1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19803 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   transaction_id           19803 non-null  object
 1   product_id               19803 non-null  object
 2   customer_id              19803 non-null  object
 3   transaction_date         19803 non-null  object
 4   online_order             19803 non-null  bool  
 5   order_status             19803 non-null  object
 6   brand                    19803 non-null  object
 7   product_line             19803 non-null  object
 8   product_class            19803 non-null  object
 9   product_size             19803 non-null  object
 10  list_price               19803 non-null  object
 11  standard_cost            19803 non-null  object
 12  product_first_sold_date  19803 non-null  object
dtypes: bool(1), object(12)
memory usage: 2.0+ MB


### Let's discover each column to make sure data inside it consistance.

In [41]:
df_T1['order_status'].unique()

array(['Approved', 'Cancelled'], dtype=object)

In [42]:
df_T1['product_line'].unique()

array(['Standard', 'Road', 'Mountain', 'Touring'], dtype=object)

In [43]:
df_T1['product_class'].unique()

array(['medium', 'low', 'high'], dtype=object)

In [44]:
df_T1['product_size'].unique()

array(['medium', 'large', 'small'], dtype=object)

### ------

### Handel Duplication
### ---

In [45]:
df_T1.duplicated().sum()

0

### No duplication

### Change data type 
### ---
- Change dataype for transaction_id, product_id, customer_id to be int
- Make datatime here consistance formate first and change datatype transaction_date to be date
- order_status, brand, product_line , product_class, product_size will change data type to be category
- list_price, standard_cost will change them to be float
- online_order to be bool

In [46]:
df_T1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19803 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype 
---  ------                   --------------  ----- 
 0   transaction_id           19803 non-null  object
 1   product_id               19803 non-null  object
 2   customer_id              19803 non-null  object
 3   transaction_date         19803 non-null  object
 4   online_order             19803 non-null  bool  
 5   order_status             19803 non-null  object
 6   brand                    19803 non-null  object
 7   product_line             19803 non-null  object
 8   product_class            19803 non-null  object
 9   product_size             19803 non-null  object
 10  list_price               19803 non-null  object
 11  standard_cost            19803 non-null  object
 12  product_first_sold_date  19803 non-null  object
dtypes: bool(1), object(12)
memory usage: 2.0+ MB


In [53]:
#change dataype for transaction_id, product_id, customer_id, product_first_sold_date to be int
df_T1['transaction_id'] = df_T1['transaction_id'].astype(int)
df_T1['product_id'] = df_T1['product_id'].astype(int)
df_T1['customer_id'] = df_T1['customer_id'].astype(int)
df_T1['product_first_sold_date'] = df_T1['product_first_sold_date'].astype(int)

# Make datatime here consistance formate first and change datatype transaction_date to be date
df_T1['transaction_date']= pd.to_datetime(df_T1['transaction_date'].astype(str),format='%Y/%m/%d')

# order_status, brand, product_line change, product_class, product_size will change data type to be category
df_T1['order_status'] = df_T1['order_status'].astype('category')
df_T1['brand'] = df_T1['brand'].astype('category')
df_T1['product_line'] = df_T1['product_line'].astype('category')
df_T1['product_class'] = df_T1['product_class'].astype('category')
df_T1['product_size'] = df_T1['product_size'].astype('category')

# list_price, standard_cost will change them to be float 
df_T1['list_price'] = df_T1['list_price'].astype('float')
df_T1['standard_cost'] = df_T1['standard_cost'].astype('float')

# check our change 
df_T1.info()

<ipython-input-53-44550dd408be>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_T1['transaction_id'] = df_T1['transaction_id'].astype(int)
<ipython-input-53-44550dd408be>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_T1['product_id'] = df_T1['product_id'].astype(int)
<ipython-input-53-44550dd408be>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pand

<class 'pandas.core.frame.DataFrame'>
Int64Index: 19803 entries, 1 to 20000
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   transaction_id           19803 non-null  int64         
 1   product_id               19803 non-null  int64         
 2   customer_id              19803 non-null  int64         
 3   transaction_date         19803 non-null  datetime64[ns]
 4   online_order             19803 non-null  bool          
 5   order_status             19803 non-null  category      
 6   brand                    19803 non-null  category      
 7   product_line             19803 non-null  category      
 8   product_class            19803 non-null  category      
 9   product_size             19803 non-null  category      
 10  list_price               19803 non-null  float64       
 11  standard_cost            19803 non-null  float64       
 12  product_first_sold_date  19803 n

<ipython-input-53-44550dd408be>:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_T1['transaction_date']= pd.to_datetime(df_T1['transaction_date'].astype(str),format='%Y/%m/%d')
<ipython-input-53-44550dd408be>:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_T1['order_status'] = df_T1['order_status'].astype('category')
<ipython-input-53-44550dd408be>:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead



In [54]:
df_T1.head(50)

0   transaction_id  product_id  customer_id transaction_date  online_order  \
1                1           2         2950       2017-02-25         False   
2                2           3         3120       2017-05-21          True   
3                3          37          402       2017-10-16         False   
4                4          88         3135       2017-08-31         False   
5                5          78          787       2017-10-01          True   
6                6          25         2339       2017-03-08          True   
7                7          22         1542       2017-04-21          True   
8                8          15         2459       2017-07-15         False   
9                9          67         1305       2017-08-10         False   
10              10          12         3262       2017-08-30          True   
11              11           5         1986       2017-01-17         False   
12              12          61         2783       2017-01-05          True   
13              13          35         1243       2017-02-26          True   
14              14          16         2717       2017-09-10         False   
15              15          12          247       2017-06-11         False   
16              16           3         2961       2017-10-10         False   
17              17          79         2426       2017-04-03         False   
18              18          33         1842       2017-06-02         False   
19              19          54         2268       2017-04-06          True   
20              20          25         3002       2017-01-28          True   
21              21          27         1582       2017-10-09         False   
22              22          37          595       2017-06-29          True   
23              23          37         2001       2017-04-08          True   
24              24          82          515       2017-10-18         False   
25              25          89         2822       2017-06-11         False   
26              26          64         2596       2017-01-10         False   
27              27          64         2666       2017-04-11          True   
28              28          19           76       2017-12-23         False   
29              29          72         3368       2017-10-13          True   
30              30          91         1173       2017-03-15         False   
31              31          88         2810       2017-09-05          True   
32              32           1         3214       2017-02-18         False   
33              33          25         1272       2017-03-20         False   
34              34          99         2003       2017-02-28          True   
35              35           0         2171       2017-08-20         False   
36              36          92         3356       2017-07-07         False   
37              37          14         1040       2017-01-09         False   
38              38           2         2916       2017-12-06         False   
39              39          12          427       2017-09-12         False   
40              40           0         2448       2017-11-28          True   
41              41          44         2570       2017-05-08          True   
42              42          79         3398       2017-05-14         False   
43              43          76          488       2017-03-17         False   
44              44          46          165       2017-12-22         False   
45              45          55         2787       2017-06-07          True   
46              46          66         3053       2017-06-20         False   
47              47          35          475       2017-05-07          True   
48              48          81         1305       2017-06-10          True   
49              49          86         1251       2017-06-17         False   
50              50          32         3103       2017-05-10         False   

0  order_status      

In [147]:
df_T1.tail(20)

0      transaction_id  product_id  customer_id transaction_date  online_order  \
19981           19981          95         3304       2017-06-07         False   
19982           19982          28         1075       2017-07-20          True   
19983           19983          11         1137       2017-03-11          True   
19984           19984          80          275       2017-07-02         False   
19985           19985          31          873       2017-11-15          True   
19986           19986          36         2918       2017-11-09          True   
19987           19987          38         2385       2017-11-01         False   
19988           19988           0           13       2017-04-05          True   
19989           19989           0          714       2017-04-27          True   
19990           19990          55          135       2017-01-27          True   
19991           19991          78          425       2017-01-26         False   
19992           19992          11         1374       2017-09-14         False   
19993           19993          13            5       2017-04-28         False   
19994           19994          77         2618       2017-12-23         False   
19995           19995           9          718       2017-05-13          True   
19996           19996          51         1018       2017-06-24          True   
19997           19997          41          127       2017-11-09          True   
19998           19998          87         2284       2017-04-14          True   
19999           19999           6         2764       2017-07-03         False   
20000           20000          11         1144       2017-09-22          True   

0     order_status           brand product_line product_class product_size  \
19981     Approved  Giant Bicycles     Standard        medium        large   
19982     Approved           Solex         Road        medium        small   
19983     Approved  Giant Bicycles     Standard          high       medium   
19984     Approved   Trek Bicycles     Standard        medium        large   
19985     Approved  Giant Bicycles     Standard        medium       medium   
19986     Approved           Solex     Standard           low       medium   
19987     Approved           Solex     Standard        medium       medium   
19988     Approved  Norco Bicycles         Road        medium       medium   
19989     Approved  Norco Bicycles     Standard           low       medium   
19990    Cancelled   Trek Bicycles         Road        medium        large   
19991     Approved  Giant Bicycles     Standard        medium        large   
19992     Approved  Giant Bicycles     Standard          high       medium   
19993     Approved           Solex     Standard        medium       medium   
19994     Approved  Norco Bicycles         Road        medium        large   
19995     Approved      OHM Cycles         Road        medium       medium   
19996     Approved      OHM Cycles     Standard          high       medium   
19997     Approved           Solex         Road        medium       medium   
19998     Approved      OHM Cycles     Standard        medium       medium   
19999     Approved      OHM Cycles     Standard          high       medium   
20000     Approved   Trek Bicycles     Standard        medium        small   

0      list_price  standard_cost  product_first_sold_date  
19981      569.56         528.43                    37874  
19982     1703.52        1516.13                    40649  
19983     1274.93         764.96                    39298  
19984     1469.44         596.55                    41047  
19985      230.91         173.18                    34527  
19986      945.04         507.58                    39526  
19987     1577.53         826.51                    40618  
19988      544.05         376.84                    42105  
19989      363.01         290.41                    37626  
19990     1894.19         598.76                    37823  

### Transaction Table now clean and ready for the next stage
- Just I need to understand some data meaing as  (product_first_sold_date)

## ----------------------------------------------------------------

## Data Pepration NewCustomerList
## ---

In [67]:
#read data set NewCustomerList
df_NCL = pd.read_excel(r'KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='NewCustomerList')
df_NCL.head(50)

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                          first_name                                                                                                                              
1                                             Chickie                                                                                                                              
2                                               Morly                                                                                                                              
3                                             Ardelis                                                                                                                              
4                                              Lucine                                                                                                                              
5                                             Melinda                                                                                                                              
6                                               Druci                                                                                                                              
7                                            Rutledge                                                                                                                              
8                                              Nancie                                                                                                                              
9                                                Duff                                                                                                                              
10                                            Barthel                                                                                                                              
11                                           Rockwell                                                                                                                              
12                                            Wheeler                                                                                                                              
13                                               Olag                                                                                                                              
14                                              Melba                                                                                                                              
15                                             Mandie                                                                                                                              
16                                              Dukie                                                                                                                              
17                                           Marcelia                                                                                                                              
18                                          Winnifred                                                                                                                              
19                                             Odilia                                                                                                                              
20                                              Karly                                                                                                                              
21                                         

### Observation NewCustomerList dataset:
- DOB not have consistance formate one row with (date) and other row with Date & time
- What's the mean with past_3_years_bike_related_purchases?
- There are missing data in last_name column
- job_title have missing data (Nan) and extra number
- Job_industery_category have missing data 
- What's the deceased_indicator?
- There are 5 colums with no title (NaN) and have decimal values

In [68]:
# remove first row and adjust header
h1= df_NCL.iloc[0]
df_NCL = df_NCL[1:]
df_NCL.columns = h1
df_NCL.head(50)

0  first_name     last_name  gender past_3_years_bike_related_purchases  \
1     Chickie       Brister    Male                                  86   
2       Morly        Genery    Male                                  69   
3     Ardelis     Forrester  Female                                  10   
4      Lucine         Stutt  Female                                  64   
5     Melinda        Hadlee  Female                                  34   
6       Druci       Brandli  Female                                  39   
7    Rutledge         Hallt    Male                                  23   
8      Nancie          Vian  Female                                  74   
9        Duff     Karlowicz    Male                                  50   
10    Barthel        Docket    Male                                  72   
11   Rockwell        Matson    Male                                  94   
12    Wheeler       Winward    Male                                  48   
13       Olag           NaN    Male                                  60   
14      Melba      Spellacy  Female                                  38   
15     Mandie        Feares  Female                                  32   
16      Dukie         Swire    Male                                  88   
17   Marcelia     Monkleigh  Female                                  61   
18  Winnifred  Beswetherick  Female                                  83   
19     Odilia         Quick  Female                                  65   
20      Karly     Willavize  Female                                   2   
21     Teddie      Burchill    Male                                  11   
22     Gaston     Dallaghan    Male                                  44   
23       Otis         Ottey    Male                                  26   
24   Tabbatha       Averill  Female                                   5   
25      Brena     Schnitter  Female                                  78   
26     Rourke      Gillbard    Male                                  11   
27      Dyane       Burwell  Female                                  19   
28   Claudine      Barstowk  Female                                  71   
29    Blinnie          Roze  Female                                  84   
30      Rhona     De Freyne  Female                                  45   
31    Sharron     Claibourn  Female                                  62   
32      Brien        Heaton    Male                                  70   
33    Sybilla       MacCart  Female                                  88   
34      Mikel        McNess    Male                                  71   
35     Maisie        Maddox  Female                                  27   
36     Arleen       Casbolt  Female                                  48   
37     Farlie       Petford    Male                                  76   
38   Mitchell      MacCague    Male                                  58   
39      Garik      Whitwell    Male                                  44   
40    Antonin         Britt    Male                                  64   
41      Vinny       Incogna  Female                                  73   
42     Colene     Fishleigh  Female                                  24   
43      Neile        Argent  Female                                  79   
44    Corinna        Suggey  Female                                  52   
45     Brooke        Arling    Male                                  76   
46      Gipsy      Ewestace  Female                                  29   
47     Sheena        Kybbye  Female                                  14   
48     Jobina       Gobourn  Female                                  85   
49       Gale       Disbrow  Female                                  59   
50    Thaxter     Kingsbury    Male                                  12   

0                   DOB                             job_title  \
1            1957-07-12                       General Manager   
2            1970-03-22                   St

In [69]:
df_NCL.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 1 to 1000
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   first_name                           1000 non-null   object 
 1   last_name                            971 non-null    object 
 2   gender                               1000 non-null   object 
 3   past_3_years_bike_related_purchases  1000 non-null   object 
 4   DOB                                  983 non-null    object 
 5   job_title                            894 non-null    object 
 6   job_industry_category                835 non-null    object 
 7   wealth_segment                       1000 non-null   object 
 8   deceased_indicator                   1000 non-null   object 
 9   owns_car                             1000 non-null   object 
 10  tenure                               1000 non-null   object 
 11  address                       

## Obsevation NewClientList Table
- Change data type for gender, job_title,job_industry_category, wealth_segment, state to be category
- Change data type for past_3_years_bike_related_purchases, tenure,postcode, property_valuation  to be int
- Adjust DOB formate for data and change data type to be datetime
- Adjust data type for owns_car to be bool
- I need to name of columns name which called nan. Becaue, it's have value and may be will be useful for our analysis.

In [70]:
# let's know how many rows 
len(df_NCL.index)

1000

In [71]:
# we will know how many missing values in that data set
for i in df_NCL.columns :
    if df_NCL[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_NCL[i].isna().sum())

Null Values in  last_name Column : 29
Null Values in  DOB Column : 17
Null Values in  job_title Column : 106
Null Values in  job_industry_category Column : 165


In [72]:
# last_name Column : 29 missing. It's essential for our analysis. So, I will fill data with 'Unknown' simple approach
df_NCL['last_name'] = df_NCL['last_name'].fillna('Unknown')



#job_title Column : 106 missing & job_industry_category Column : 165. 
# Simple approach, I will fill data with 'Unknown' simple approach.
# If this columns, is essential for our analysis. So, we can use NLP techniques to extract job title from other text columns or try to impute them based on pattern within dataset
df_NCL['job_title'] = df_NCL['job_title'].fillna('Unknown')
df_NCL['job_industry_category'] = df_NCL['job_industry_category'].fillna('Unknown')

# check there is no missing data 
for i in df_NCL.columns :
    if df_NCL[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_NCL[i].isna().sum())

Null Values in  DOB Column : 17


In [77]:
# DOB Column : 17 missing. I choose to drop 17 missing, especially if the missing values is relatively small
df_NCL =df_NCL.dropna(subset = ['DOB'])

# check there is no missing data 
for i in df_NCL.columns :
    if df_NCL[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_NCL[i].isna().sum())

In [80]:
# will drop the all colums with nan values 
df_NCL= df_NCL.dropna(how="any", axis=1)

#Check dataset 
df_NCL.head(10)

0  first_name  last_name  gender past_3_years_bike_related_purchases  \
1     Chickie    Brister    Male                                  86   
2       Morly     Genery    Male                                  69   
3     Ardelis  Forrester  Female                                  10   
4      Lucine      Stutt  Female                                  64   
5     Melinda     Hadlee  Female                                  34   
6       Druci    Brandli  Female                                  39   
7    Rutledge      Hallt    Male                                  23   
8      Nancie       Vian  Female                                  74   
9        Duff  Karlowicz    Male                                  50   
10    Barthel     Docket    Male                                  72   

0                   DOB                     job_title job_industry_category  \
1            1957-07-12               General Manager         Manufacturing   
2            1970-03-22           Structural Engineer              Property   
3   1974-08-28 00:00:00        Senior Cost Accountant    Financial Services   
4            1979-01-28    Account Representative III         Manufacturing   
5            1965-09-21             Financial Analyst    Financial Services   
6            1951-04-29       Assistant Media Planner         Entertainment   
7            1976-10-06          Compensation Analyst    Financial Services   
8            1972-12-27  Human Resources Assistant II                Retail   
9            1972-04-28            Speech Pathologist         Manufacturing   
10           1985-08-02       Accounting Assistant IV                    IT   

0      wealth_segment deceased_indicator owns_car tenure  \
1       Mass Customer                  N      Yes     14   
2       Mass Customer                  N       No     16   
3   Affluent Customer                  N       No     10   
4   Affluent Customer                  N      Yes      5   
5   Affluent Customer                  N       No     19   
6      High Net Worth                  N      Yes     22   
7       Mass Customer                  N       No      8   
8       Mass Customer                  N      Yes     10   
9       Mass Customer                  N      Yes      5   
10      Mass Customer                  N      Yes     17   

0                 address postcode state    country property_valuation   NaN  \
1        45 Shopko Center     4500   QLD  Australia                  6  0.56   
2       14 Mccormick Park     2113   NSW  Australia                 11  0.89   
3     5 Colorado Crossing     3505   VIC  Australia                  5  1.01   
4      207 Annamark Plaza     4814   QLD  Australia                  1  0.87   
5       115 Montana Place     2093   NSW  Australia                  9  0.52   
6   89105 Pearson Terrace     4075   QLD  Australia                  7  0.43   
7       7 Nevada Crossing     2620   NSW  Australia                  7  0.40   
8        85 Carioca Point     4814   QLD  Australia                  5  0.58   
9          717 West Drive     2200   NSW  Australia                 10  1.03   
10   80 Scofield Junction     4151   QLD  Australia                  5  0.84   

0      NaN       NaN       NaN   NaN Rank     Value  
1   0.7000  0.875000  0.743750   1.0    1   1.71875  
2   0.8900  1.112500  0.945625   1.0    1   1.71875  
3   1.0100  1.010000  1.010000   1.0    1   1.71875  
4   1.0875  1.087500  1.087500   4.0    4  1.703125  
5   0.5200  0.650000  0.650000   4.0    4  1.703125  
6   0.5375  0.537500  0.537500   6.0    6  1.671875  
7   0.4000  0.400000  0.340000   6.0    6  1.671875  
8   0.7250  0.725000  0.616250   8.0    8   1.65625  
9   1.2875  1.609375  1.367969   8.0    8   1.65625  
10  1.0500  1.050000  0.892500  10.0   10  1.640625

### Handel duplication
### ------

In [78]:
df_NCL.duplicated().sum()

0

#### No duplication 

### Change data type 
### ---
- Change data type for gender, job_title,job_industry_category, wealth_segment, state to be category
- Change data type for past_3_years_bike_related_purchases, tenure,postcode, property_valuation  to be int
- Adjust DOB formate for data and change data type to be datetime
- Adjust data type for owns_car to be bool

In [79]:
df_NCL.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 983 entries, 1 to 1000
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   first_name                           983 non-null    object 
 1   last_name                            983 non-null    object 
 2   gender                               983 non-null    object 
 3   past_3_years_bike_related_purchases  983 non-null    object 
 4   DOB                                  983 non-null    object 
 5   job_title                            983 non-null    object 
 6   job_industry_category                983 non-null    object 
 7   wealth_segment                       983 non-null    object 
 8   deceased_indicator                   983 non-null    object 
 9   owns_car                             983 non-null    object 
 10  tenure                               983 non-null    object 
 11  address                        

In [82]:

# Change data type for gender, job_title,job_industry_category, wealth_segment, state to be category
df_NCL['gender'] = df_NCL['gender'].astype('category')
df_NCL['job_title'] = df_NCL['job_title'].astype('category')
df_NCL['job_industry_category'] = df_NCL['job_industry_category'].astype('category')
df_NCL['wealth_segment'] = df_NCL['wealth_segment'].astype('category')
df_NCL['state'] = df_NCL['state'].astype('category')

#Change data type for past_3_years_bike_related_purchases, tenure,postcode, property_valuation  to be int
df_NCL['past_3_years_bike_related_purchases'] = df_NCL['past_3_years_bike_related_purchases'].astype(int)
df_NCL['tenure'] = df_NCL['tenure'].astype(int)
df_NCL['postcode'] = df_NCL['postcode'].astype(int)
df_NCL['property_valuation'] = df_NCL['property_valuation'].astype(int)

#Adjust DOB formate for data and change data type to be datetime
df_NCL['DOB']= pd.to_datetime(df_NCL['DOB'].astype(str),format='%Y/%m/%d')

#Adjust data type for owns_car to be bool
df_NCL['owns_car'] = df_NCL['owns_car'].astype('bool')

# check our change 
df_NCL.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 983 entries, 1 to 1000
Data columns (total 23 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   first_name                           983 non-null    object        
 1   last_name                            983 non-null    object        
 2   gender                               983 non-null    category      
 3   past_3_years_bike_related_purchases  983 non-null    int64         
 4   DOB                                  983 non-null    datetime64[ns]
 5   job_title                            983 non-null    category      
 6   job_industry_category                983 non-null    category      
 7   wealth_segment                       983 non-null    category      
 8   deceased_indicator                   983 non-null    object        
 9   owns_car                             983 non-null    bool          
 10  tenure       

### Make sure quality of data for each column
### -----

In [83]:
df_NCL['gender'].unique()

['Male', 'Female']
Categories (2, object): ['Male', 'Female']

In [85]:
df_NCL['job_title'].unique().tolist()

['General Manager',
 'Structural Engineer',
 'Senior Cost Accountant',
 'Account Representative III',
 'Financial Analyst',
 'Assistant Media Planner',
 'Compensation Analyst',
 'Human Resources Assistant II',
 'Speech Pathologist',
 'Accounting Assistant IV',
 'Programmer Analyst I',
 'Environmental Specialist',
 'Human Resources Manager',
 'VP Marketing',
 'Clinical Specialist',
 'Unknown',
 'Associate Professor',
 'Actuary',
 'Internal Auditor',
 'Programmer I',
 'Quality Engineer',
 'Quality Control Specialist',
 'Account Executive',
 'Engineer II',
 'Nurse',
 'Librarian',
 'Paralegal',
 'Assistant Manager',
 'Senior Quality Engineer',
 'Recruiting Manager',
 'Design Engineer',
 'Business Systems Development Analyst',
 'VP Quality Control',
 'Cost Accountant',
 'Product Engineer',
 'Social Worker',
 'Payment Adjustment Coordinator',
 'Marketing Assistant',
 'Recruiter',
 'VP Product Management',
 'Media Manager III',
 'Developer I',
 'Senior Financial Analyst',
 'Electrical Enginee

In [90]:
# My observation regard this column, there are redundancy. We need to reomve I,II,III, IV, V, p  

# Replace pattern of string using regular expression.
df_NCL['job_title'] = df_NCL['job_title'].str.replace('I', '', regex=True)
df_NCL['job_title'] = df_NCL['job_title'].str.replace('II', '', regex=True)
df_NCL['job_title'] = df_NCL['job_title'].str.replace('III', '', regex=True)
df_NCL['job_title'] = df_NCL['job_title'].str.replace('IV', '', regex=True)
df_NCL['job_title'] = df_NCL['job_title'].str.replace('V', '', regex=True)
df_NCL['job_title'] = df_NCL['job_title'].str.replace('P ', '', regex=True)
df_NCL['job_title'].unique().tolist()

['General Manager',
 'Structural Engineer',
 'Senior Cost Accountant',
 'Account Representative ',
 'Financial Analyst',
 'Assistant Media Planner',
 'Compensation Analyst',
 'Human Resources Assistant ',
 'Speech Pathologist',
 'Accounting Assistant ',
 'Programmer Analyst ',
 'Environmental Specialist',
 'Human Resources Manager',
 'Marketing',
 'Clinical Specialist',
 'Unknown',
 'Associate Professor',
 'Actuary',
 'nternal Auditor',
 'Programmer ',
 'Quality Engineer',
 'Quality Control Specialist',
 'Account Executive',
 'Engineer ',
 'Nurse',
 'Librarian',
 'Paralegal',
 'Assistant Manager',
 'Senior Quality Engineer',
 'Recruiting Manager',
 'Design Engineer',
 'Business Systems Development Analyst',
 'Quality Control',
 'Cost Accountant',
 'Product Engineer',
 'Social Worker',
 'Payment Adjustment Coordinator',
 'Marketing Assistant',
 'Recruiter',
 'Product Management',
 'Media Manager ',
 'Developer ',
 'Senior Financial Analyst',
 'Electrical Engineer',
 'Accountant ',
 'Exe

In [92]:
df_NCL['job_industry_category'].unique().tolist()

['Manufacturing',
 'Property',
 'Financial Services',
 'Entertainment',
 'Retail',
 'IT',
 'Telecommunications',
 'Health',
 'Unknown',
 'Argiculture']

In [93]:
df_NCL['wealth_segment'].unique()

['Mass Customer', 'Affluent Customer', 'High Net Worth']
Categories (3, object): ['Mass Customer', 'Affluent Customer', 'High Net Worth']

In [94]:
df_NCL['owns_car'].unique()

array([ True])

In [95]:
df_NCL['address'].unique()

array(['45 Shopko Center', '14 Mccormick Park', '5 Colorado Crossing',
       '207 Annamark Plaza', '115 Montana Place', '89105 Pearson Terrace',
       '7 Nevada Crossing', '85 Carioca Point', '717 West Drive',
       '80 Scofield Junction', '3682 Crowley Point',
       '3 Golden Leaf Point', '0484 North Avenue', '0591 Anzinger Circle',
       '39 Kedzie Pass', '64 Granby Parkway', '610 Swallow Street',
       '61 4th Street', '1550 Russell Way', '193 North Point',
       '321 Raven Plaza', '656 Fuller Street', '1562 Merchant Street',
       '663 8th Parkway', '67 Shelley Street', '75 Cordelia Trail',
       '51 Hooker Court', '1859 Forest Circle', '44557 Rutledge Court',
       '11184 East Drive', '555 Hermina Avenue', '8 Novick Trail',
       '74 Welch Pass', '3 Pleasure Drive', '8 Dennis Point',
       '41042 Lotheville Crossing', '2330 Butternut Trail',
       '240 Acker Avenue', '04 Dexter Way', '011 Northland Trail',
       '8 Grayhawk Circle', '44 Darwin Lane', '2548 Arrowood P

In [96]:
df_NCL['state'].unique()

['QLD', 'NSW', 'VIC']
Categories (3, object): ['QLD', 'NSW', 'VIC']

In [97]:
df_NCL['country'].unique()

array(['Australia'], dtype=object)

### The NewClientList Table now have quality data. Just want to understand some meaning for colums and name of nan colums to make ready for the nexr amalysis 

### ------------------------------------------------------------------------------

## CustomerDemographic table
## ----

In [110]:
# dataset for CustomerDemographic
df_CD = pd.read_excel(r'KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerDemographic')
df_CD.head(50)

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                         customer_id                                                                                                                              
1                                                   1                                                                                                                              
2                                                   2                                                                                                                              
3                                                   3                                                                                                                              
4                                                   4                                                                                                                              
5                                                   5                                                                                                                              
6                                                   6                                                                                                                              
7                                                   7                                                                                                                              
8                                                   8                                                                                                                              
9                                                   9                                                                                                                              
10                                                 10                                                                                                                              
11                                                 11                                                                                                                              
12                                                 12                                                                                                                              
13                                                 13                                                                                                                              
14                                                 14                                                                                                                              
15                                                 15                                                                                                                              
16                                                 16                                                                                                                              
17                                                 17                                                                                                                              
18                                                 18                                                                                                                              
19                                                 19                                                                                                                              
20                                                 20                                                                                                                              
21                                         

### Observation CustomerDemographic dataset:
- Gender do not have the same format  sometime use F/M sometime use Male / Female 
- Gender have wrong data, as irrelevant data as(U)
- DOB do not have the same formate
- job_title have missing data 
- job_industry_category have missing data 
- default column have understanduable data. what the porpuse of it.
- There 13 column NaN without any data.''

In [111]:
# let's know how many rows 
len(df_CD.index)

4001

In [112]:
#change header
h1 =df_CD.iloc[0]
df_CD = df_CD[1:]
df_CD.columns= h1
df_CD.head(20)

0  customer_id      first_name  last_name  gender  \
1            1         Laraine  Medendorp       F   
2            2             Eli    Bockman    Male   
3            3           Arlin     Dearle    Male   
4            4          Talbot        NaN    Male   
5            5  Sheila-kathryn     Calton  Female   
6            6            Curr  Duckhouse    Male   
7            7            Fina     Merali  Female   
8            8             Rod      Inder    Male   
9            9            Mala       Lind  Female   
10          10        Fiorenze    Birdall  Female   
11          11           Uriah     Bisatt    Male   
12          12         Sawyere   Flattman    Male   
13          13        Gabriele   Norcross    Male   
14          14        Rayshell   Kitteman  Female   
15          15          Erroll     Radage    Male   
16          16          Harlin       Parr    Male   
17          17           Heath    Faraday    Male   
18          18          Marjie    Neasham  Female   
19          19          Sorcha     Keyson  Female   
20          20          Basile      Firth    Male   

0  past_3_years_bike_related_purchases                  DOB  \
1                                   93  1953-10-12 00:00:00   
2                                   81  1980-12-16 00:00:00   
3                                   61  1954-01-20 00:00:00   
4                                   33  1961-10-03 00:00:00   
5                                   56  1977-05-13 00:00:00   
6                                   35  1966-09-16 00:00:00   
7                                    6  1976-02-23 00:00:00   
8                                   31  1962-03-30 00:00:00   
9                                   97  1973-03-10 00:00:00   
10                                  49  1988-10-11 00:00:00   
11                                  99  1954-04-30 00:00:00   
12                                  58  1994-07-21 00:00:00   
13                                  38  1955-02-15 00:00:00   
14                                  85  1983-03-25 00:00:00   
15                                  91  2000-07-13 00:00:00   
16                                  38  1977-02-27 00:00:00   
17                                  57  1962-03-19 00:00:00   
18                                  79  1967-07-06 00:00:00   
19                                  76  2001-04-15 00:00:00   
20                                  72  1980-08-13 00:00:00   

0                              job_title job_industry_category  \
1                    Executive Secretary                Health   
2                 Administrative Officer    Financial Services   
3                     Recruiting Manager              Property   
4                                    NaN                    IT   
5                          Senior Editor                   NaN   
6                                    NaN                Retail   
7                                    NaN    Financial Services   
8                        Media Manager I                   NaN   
9   Business Systems Development Analyst           Argiculture   
10               Senior Quality Engineer    Financial Services   
11                                   NaN              Property   
12                Nuclear Power Engineer         Manufacturing   
13                           Developer I    Financial Services   
14                     Account Executive    Financial Services   
15                      Junior Executive         Manufacturing   
16                      Media Manager IV                   NaN   
17                       Sales Associate                   NaN   
18                             Professor                   NaN   
19                   Geological Engineer         Manufacturing   
20                       Project Manager         Manufacturing   

0      wealth_segment deceased_indicator  \
1       Mass Customer                  N   
2       Mass Customer                  N   
3       Mass Customer                  N   

In [113]:
df_CD.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4000 entries, 1 to 4000
Data columns (total 26 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   customer_id                          4000 non-null   object 
 1   first_name                           4000 non-null   object 
 2   last_name                            3875 non-null   object 
 3   gender                               4000 non-null   object 
 4   past_3_years_bike_related_purchases  4000 non-null   object 
 5   DOB                                  3913 non-null   object 
 6   job_title                            3494 non-null   object 
 7   job_industry_category                3344 non-null   object 
 8   wealth_segment                       4000 non-null   object 
 9   deceased_indicator                   4000 non-null   object 
 10  default                              3698 non-null   object 
 11  owns_car                      

## Obsevation CustomerDemographic Table
- Adjust customer_id, past_3_years_bike_related_purchases data type to be int
- gender column we need to make data formate consistance then change data type to be category and i suggest to be (Female / Male) to make sure data the same. since NewClientList table have the same column in that fomrate.
- Change DOB data type to be datetime
- Adjust data type of job_title,job_industry_category, wealth_segment to be category
- Drop nan colums

In [114]:
# we will know how many missing values in that data set
for i in df_CD.columns :
    if df_CD[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_CD[i].isna().sum())

Null Values in  last_name Column : 125
Null Values in  DOB Column : 87
Null Values in  job_title Column : 506
Null Values in  job_industry_category Column : 656
Null Values in  default Column : 302
Null Values in  tenure Column : 87
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000
Null Values in  nan Column : 4000


In [115]:
len(df_CD.index)

4000

### Handel missing data 
### -------
-  last_name Column : 125
-  DOB Column : 87
-  job_title Column : 506
-  job_industry_category Column : 656
-  default Column : 302
-  tenure Column : 87
-  colums with nan data

In [116]:
# last_name Column : 125 missing. It's essential for our analysis. So, I will fill data with 'Unknown' simple approach
df_CD['last_name'] = df_CD['last_name'].fillna('Unknown')



#job_title Column : 506 missing & job_industry_category Column : 656. 
# Simple approach, I will fill data with 'Unknown' simple approach.
# If this columns, is essential for our analysis. So, we can use NLP techiques to extract job title from other text columns or try to impute them based on pattern within dataset
df_CD['job_title'] = df_CD['job_title'].fillna('Unknown')
df_CD['job_industry_category'] = df_CD['job_industry_category'].fillna('Unknown')

#DOB Column : 87 missing. drop data
df_CD =df_CD.dropna(subset = ['DOB'])

# will drop the all columns with nan values 
df_CD= df_CD.dropna(how="any", axis=1)

# check there is no missing data 
for i in df_CD.columns :
    if df_CD[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_CD[i].isna().sum())

## Check duplication
## ---

In [107]:
df_CD.duplicated().sum()

0

### No duplication

## Change data type
## ---
- Adjust customer_id, past_3_years_bike_related_purchases data type to be int
- gender column we need to make data formate consistance then change data type to be category and i suggest to be (Female / Male) to make sure data the same. since NewClientList table have the same column in that fomrate.
- Adjust data type of job_title,job_industry_category, wealth_segment to be category
- Change DOB data type to be datetime
- Change owns_car data type to be bool


In [123]:
#Adjust customer_id, past_3_years_bike_related_purchases data type to be int
df_CD['past_3_years_bike_related_purchases'] = df_CD['past_3_years_bike_related_purchases'].astype(int)
df_CD['customer_id'] = df_CD['customer_id'].astype(int)

# Change data type for gender, job_title,job_industry_category, wealth_segment to be category
df_CD['gender'] = df_CD['gender'].astype('category')
df_CD['job_title'] = df_CD['job_title'].astype('category')
df_CD['job_industry_category'] = df_CD['job_industry_category'].astype('category')
df_CD['wealth_segment'] = df_CD['wealth_segment'].astype('category')

#Adjust DOB formate for data and change data type to be datetime
df_CD['DOB']= pd.to_datetime(df_CD['DOB'].astype(str),format='%Y/%m/%d')

#Change owns_car data type to be bool
df_CD['owns_car']= df_CD['owns_car'].astype('bool')

# check our change 
df_CD.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3912 entries, 1 to 4000
Data columns (total 12 columns):
 #   Column                               Non-Null Count  Dtype         
---  ------                               --------------  -----         
 0   customer_id                          3912 non-null   int64         
 1   first_name                           3912 non-null   object        
 2   last_name                            3912 non-null   object        
 3   gender                               3912 non-null   category      
 4   past_3_years_bike_related_purchases  3912 non-null   int64         
 5   DOB                                  3912 non-null   datetime64[ns]
 6   job_title                            3912 non-null   category      
 7   job_industry_category                3912 non-null   category      
 8   wealth_segment                       3912 non-null   category      
 9   deceased_indicator                   3912 non-null   object        
 10  owns_car    

### Check quality of data 
### ----
- gender                                   
- job_title                                 
- job_industry_category                    
- wealth_segment                            
- deceased_indicator                           
- owns_car      

In [118]:
df_CD['gender'].unique()

['F', 'Male', 'Female', 'U', 'Femal', 'M']
Categories (6, object): ['F', 'Male', 'Female', 'U', 'Femal', 'M']

In [120]:
# replace {F , Femal} to be Female
df_CD['gender'] = df_CD['gender'].replace('F', 'Female')
df_CD['gender'] = df_CD['gender'].replace('Femal', 'Female')

# replace {M } to be  Male
df_CD['gender'] = df_CD['gender'].replace('M', 'Male')

# check replace
df_CD['gender'].unique()

['Female', 'Male', 'U']
Categories (3, object): ['Female', 'Male', 'U']

In [121]:
# group by gender to know how many 'U' it's have 
df_CD.groupby('gender').size()

gender
Female    2039
Male      1873
U            1
dtype: int64

In [ ]:
# we can just drop gender with 'U' value 

In [122]:
df_CD =df_CD[(df_CD.gender != 'U')]
# confirm replace 
df_CD.gender.unique()

['Female', 'Male']
Categories (2, object): ['Female', 'Male']

In [125]:
# check job_title
df_CD['job_title'].unique().tolist()

['Executive Secretary',
 'Administrative Officer',
 'Recruiting Manager',
 'Unknown',
 'Senior Editor',
 'Media Manager I',
 'Business Systems Development Analyst',
 'Senior Quality Engineer',
 'Nuclear Power Engineer',
 'Developer I',
 'Account Executive',
 'Junior Executive',
 'Media Manager IV',
 'Sales Associate',
 'Professor',
 'Geological Engineer',
 'Project Manager',
 'Safety Technician I',
 'Research Assistant I',
 'Accounting Assistant III',
 'Editor',
 'Research Nurse',
 'Safety Technician III',
 'Staff Accountant III',
 'Legal Assistant',
 'Product Engineer',
 'Information Systems Manager',
 'VP Quality Control',
 'Social Worker',
 'Senior Cost Accountant',
 'Assistant Media Planner',
 'Payment Adjustment Coordinator',
 'Food Chemist',
 'Accountant III',
 'Director of Sales',
 'Senior Financial Analyst',
 'Registered Nurse',
 'Biostatistician II',
 'Computer Systems Analyst II',
 'Software Test Engineer II',
 'Paralegal',
 'VP Sales',
 'Chief Design Engineer',
 'Office Assi

In [127]:
# My observation regard this column, there are redundancy. We need to reomve I,II,III, IV, V, P  

# Replace pattern of string using regular expression.
df_CD['job_title'] = df_CD['job_title'].str.replace('I', '', regex=True)
df_CD['job_title'] = df_CD['job_title'].str.replace('II', '', regex=True)
df_CD['job_title'] = df_CD['job_title'].str.replace('III', '', regex=True)
df_CD['job_title'] = df_CD['job_title'].str.replace('IV', '', regex=True)
df_CD['job_title'] = df_CD['job_title'].str.replace('V', '', regex=True)
df_CD['job_title'] = df_CD['job_title'].str.replace('P ', '', regex=True)

df_CD['job_title'].unique().tolist()

['Executive Secretary',
 'Administrative Officer',
 'Recruiting Manager',
 'Unknown',
 'Senior Editor',
 'Media Manager ',
 'Business Systems Development Analyst',
 'Senior Quality Engineer',
 'Nuclear Power Engineer',
 'Developer ',
 'Account Executive',
 'Junior Executive',
 'Sales Associate',
 'Professor',
 'Geological Engineer',
 'Project Manager',
 'Safety Technician ',
 'Research Assistant ',
 'Accounting Assistant ',
 'Editor',
 'Research Nurse',
 'Staff Accountant ',
 'Legal Assistant',
 'Product Engineer',
 'nformation Systems Manager',
 'Quality Control',
 'Social Worker',
 'Senior Cost Accountant',
 'Assistant Media Planner',
 'Payment Adjustment Coordinator',
 'Food Chemist',
 'Accountant ',
 'Director of Sales',
 'Senior Financial Analyst',
 'Registered Nurse',
 'Biostatistician ',
 'Computer Systems Analyst ',
 'Software Test Engineer ',
 'Paralegal',
 'Sales',
 'Chief Design Engineer',
 'Office Assistant ',
 'Physical Therapy Assistant',
 'Help Desk Operator',
 'Web Deve

In [129]:
# check job_industry_category
df_CD['job_industry_category'].unique().tolist()

['Health',
 'Financial Services',
 'Property',
 'IT',
 'Unknown',
 'Retail',
 'Argiculture',
 'Manufacturing',
 'Telecommunications',
 'Entertainment']

In [130]:
# check wealth_segment
df_CD['wealth_segment'].unique()

['Mass Customer', 'Affluent Customer', 'High Net Worth']
Categories (3, object): ['Mass Customer', 'Affluent Customer', 'High Net Worth']

In [131]:
# check deceased_indicator
df_CD['deceased_indicator'].unique()

array(['N', 'Y'], dtype=object)

In [133]:
# check owns_car
df_CD['owns_car'].unique()

array([ True])

### -------------------------------------------------------------------

## CustomerAddress table
## ----------------

In [134]:
#read data set CustomerAddress
df_CA = pd.read_excel(r'KPMG_VI_New_raw_data_update_final.xlsx', sheet_name='CustomerAddress')
df_CA.head(50)

Note: The data and information in this document is reflective of a hypothetical situation and client. This document is to be used for KPMG Virtual Internship purposes only.   \
0                                         customer_id                                                                                                                              
1                                                   1                                                                                                                              
2                                                   2                                                                                                                              
3                                                   4                                                                                                                              
4                                                   5                                                                                                                              
5                                                   6                                                                                                                              
6                                                   7                                                                                                                              
7                                                   8                                                                                                                              
8                                                   9                                                                                                                              
9                                                  11                                                                                                                              
10                                                 12                                                                                                                              
11                                                 13                                                                                                                              
12                                                 14                                                                                                                              
13                                                 15                                                                                                                              
14                                                 16                                                                                                                              
15                                                 17                                                                                                                              
16                                                 18                                                                                                                              
17                                                 19                                                                                                                              
18                                                 20                                                                                                                              
19                                                 21                                                                                                                              
20                                                 24                                                                                                                              
21                                         

### Observation CustomerAddress dataset:
- 20 NaN colums 

In [135]:
h1 = df_CA.iloc[0]
df_CA = df_CA[1:]
df_CA.columns = h1
df_CA.head(50)

0  customer_id                   address postcode            state    country  \
1            1        060 Morning Avenue     2016  New South Wales  Australia   
2            2       6 Meadow Vale Court     2153  New South Wales  Australia   
3            4        0 Holy Cross Court     4211              QLD  Australia   
4            5       17979 Del Mar Point     2448  New South Wales  Australia   
5            6          9 Oakridge Court     3216              VIC  Australia   
6            7          4 Delaware Trail     2210  New South Wales  Australia   
7            8       49 Londonderry Lane     2650  New South Wales  Australia   
8            9           97736 7th Trail     2023  New South Wales  Australia   
9           11      93405 Ludington Park     3044              VIC  Australia   
10          12   44339 Golden Leaf Alley     4557              QLD  Australia   
11          13       2 Sutherland Street     3799              VIC  Australia   
12          14           9 Mcbride Trail     2760  New South Wales  Australia   
13          15    9861 New Castle Avenue     2428  New South Wales  Australia   
14          16          52 Moland Street     3331              VIC  Australia   
15          17     82391 Kensington Lane     3058              VIC  Australia   
16          18             092 2nd Alley     2135  New South Wales  Australia   
17          19         59 Spaight Circle     2233  New South Wales  Australia   
18          20      032 Bartelt Crossing     2444  New South Wales  Australia   
19          21           18 Jenna Center     2650  New South Wales  Australia   
20          24          3 Cordelia Plaza     2153  New South Wales  Australia   
21          25             28 5th Center     4413              QLD  Australia   
22          26            52 Carey Alley     4740              QLD  Australia   
23          27            96 Texas Plaza     3218              VIC  Australia   
24          28           48 Eagan Avenue     4868              QLD  Australia   
25          29              9 Buell Park     4116              QLD  Australia   
26          30       02663 Buell Parkway     2519  New South Wales  Australia   
27          31       2294 Pleasure Place     2135  New South Wales  Australia   
28          32       2951 Petterle Place     2756  New South Wales  Australia   
29          33           63 Lukken Drive     2170  New South Wales  Australia   
30          34            833 Luster Way     4005              QLD  Australia   
31          35          34 Russell Plaza     4127              QLD  Australia   
32          36           78 Del Sol Hill     2535  New South Wales  Australia   
33          37             54 Sage Plaza     4726              QLD  Australia   
34          38           90 Lawn Parkway     4805              QLD  Australia   
35          39      38726 Ilene Crossing     2212  New South Wales  Australia   
36          40          3496 Brown Drive     2207  New South Wales  Australia   
37          41             7188 Cody Way     3350              VIC  Australia   
38          42          1 Claremont Park     3931              VIC  Australia   
39          43           60 Eagan Circle     4018              QLD  Australia   
40          44  89314 Eagle Crest Center     4670              QLD  Australia   
41          45       2028 Lakewood Place     2480  New South Wales  Australia   
42          46        70875 Hudson Point     3212              VIC  Australia   
43          47         543 Killdeer Park     2322  New South Wales  Australia   
44          48        22125 Ramsey Drive     3198              VIC  Australia   
45          49      3107 Calypso Terrace     2210  New South Wales  Australia   
46          50        957 Veith Junction     2047  New South Wales  Australia   
47          51       4853 Gulseth Avenue     4503              QLD  Australia   
48          52        9 Michigan Terrace     2145  New South Wales  Australia   
49          53        62 Steens

In [136]:
df_CA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 1 to 3999
Data columns (total 26 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   customer_id         3999 non-null   object 
 1   address             3999 non-null   object 
 2   postcode            3999 non-null   object 
 3   state               3999 non-null   object 
 4   country             3999 non-null   object 
 5   property_valuation  3999 non-null   object 
 6   nan                 0 non-null      float64
 7   nan                 0 non-null      float64
 8   nan                 0 non-null      float64
 9   nan                 0 non-null      float64
 10  nan                 0 non-null      float64
 11  nan                 0 non-null      float64
 12  nan                 0 non-null      float64
 13  nan                 0 non-null      float64
 14  nan                 0 non-null      float64
 15  nan                 0 non-null      float64
 16  nan   

## Obsevation CustomerAddress Table
- postcode & property_valuation change data type to be int
- state & country change data type to be categorical
- Delete nan colums

In [137]:
# let's know how many rows 
len(df_CA.index)

3999

In [138]:
# we will know how many missing values in that data set
for i in df_CA.columns :
    if df_CA[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_CA[i].isna().sum())

Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999
Null Values in  nan Column : 3999


In [140]:
# will drop the all colums with nan values 
df_CA= df_CA.dropna(how="any", axis=1)

# check there is no missing data 
for i in df_CA.columns :
    if df_CA[i].isna().sum() != 0 : 
        print('Null Values in ', i, "Column :", df_CA[i].isna().sum())

In [139]:
# Check duplication
df_CA.duplicated().sum()

0

no duplication

### Adjust data type 
### ----
- Change customer_id , postcode, property_valuation data type to be int
- Change state , country to category

In [141]:
df_CA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 1 to 3999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   customer_id         3999 non-null   object
 1   address             3999 non-null   object
 2   postcode            3999 non-null   object
 3   state               3999 non-null   object
 4   country             3999 non-null   object
 5   property_valuation  3999 non-null   object
dtypes: object(6)
memory usage: 187.6+ KB


In [142]:
# Change customer_id , postcode, property_valuation data type to be int
df_CA['customer_id'] = df_CA['customer_id'].astype(int)
df_CA['postcode'] = df_CA['postcode'].astype(int)
df_CA['property_valuation'] = df_CA['property_valuation'].astype(int)

# Change state , country to category
df_CA['state'] = df_CA['state'].astype('category')
df_CA['country'] = df_CA['country'].astype('category')

# check 
df_CA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3999 entries, 1 to 3999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   customer_id         3999 non-null   int64   
 1   address             3999 non-null   object  
 2   postcode            3999 non-null   int64   
 3   state               3999 non-null   category
 4   country             3999 non-null   category
 5   property_valuation  3999 non-null   int64   
dtypes: category(2), int64(3), object(1)
memory usage: 133.2+ KB


### Check the quality of data
### ----

In [143]:
# check state
df_CA['state'].unique()


['New South Wales', 'QLD', 'VIC', 'NSW', 'Victoria']
Categories (5, object): ['New South Wales', 'QLD', 'VIC', 'NSW', 'Victoria']

In [144]:
# check country
df_CA['country'].unique()

['Australia']
Categories (1, object): ['Australia']

In [146]:
# final data set
df_CA.head(50)

0   customer_id                   address  postcode            state  \
1             1        060 Morning Avenue      2016  New South Wales   
2             2       6 Meadow Vale Court      2153  New South Wales   
3             4        0 Holy Cross Court      4211              QLD   
4             5       17979 Del Mar Point      2448  New South Wales   
5             6          9 Oakridge Court      3216              VIC   
6             7          4 Delaware Trail      2210  New South Wales   
7             8       49 Londonderry Lane      2650  New South Wales   
8             9           97736 7th Trail      2023  New South Wales   
9            11      93405 Ludington Park      3044              VIC   
10           12   44339 Golden Leaf Alley      4557              QLD   
11           13       2 Sutherland Street      3799              VIC   
12           14           9 Mcbride Trail      2760  New South Wales   
13           15    9861 New Castle Avenue      2428  New South Wales   
14           16          52 Moland Street      3331              VIC   
15           17     82391 Kensington Lane      3058              VIC   
16           18             092 2nd Alley      2135  New South Wales   
17           19         59 Spaight Circle      2233  New South Wales   
18           20      032 Bartelt Crossing      2444  New South Wales   
19           21           18 Jenna Center      2650  New South Wales   
20           24          3 Cordelia Plaza      2153  New South Wales   
21           25             28 5th Center      4413              QLD   
22           26            52 Carey Alley      4740              QLD   
23           27            96 Texas Plaza      3218              VIC   
24           28           48 Eagan Avenue      4868              QLD   
25           29              9 Buell Park      4116              QLD   
26           30       02663 Buell Parkway      2519  New South Wales   
27           31       2294 Pleasure Place      2135  New South Wales   
28           32       2951 Petterle Place      2756  New South Wales   
29           33           63 Lukken Drive      2170  New South Wales   
30           34            833 Luster Way      4005              QLD   
31           35          34 Russell Plaza      4127              QLD   
32           36           78 Del Sol Hill      2535  New South Wales   
33           37             54 Sage Plaza      4726              QLD   
34           38           90 Lawn Parkway      4805              QLD   
35           39      38726 Ilene Crossing      2212  New South Wales   
36           40          3496 Brown Drive      2207  New South Wales   
37           41             7188 Cody Way      3350              VIC   
38           42          1 Claremont Park      3931              VIC   
39           43           60 Eagan Circle      4018              QLD   
40           44  89314 Eagle Crest Center      4670              QLD   
41           45       2028 Lakewood Place      2480  New South Wales   
42           46        70875 Hudson Point      3212              VIC   
43           47         543 Killdeer Park      2322  New South Wales   
44           48        22125 Ramsey Drive      3198              VIC   
45           49      3107 Calypso Terrace      2210  New South Wales   
46           50        957 Veith Junction      2047  New South Wales   
47           51       4853 Gulseth Avenue      4503              QLD   
48           52        9 Michigan Terrace      2145  New South Wales   
49           53        62 Steensland Road      4350              QLD   
50           54         6 Homewood Avenue      2142  New South Wales   

0     country  property_valuation  
1   Australia                  10  
2   Australia                  10  
3   Australia                   9  
4   Australia                   4  
5   Australia                   9  
6   Australia                   9  
7   Australia                   4  
8   Australia                  12  
9  